Introduce Apache Tika extension

sberyozkin opened the pull request.

sberyozkin committed:
Introduce Apache Tika extension

sberyozkin commented:
@gsmet FYI. I'll keep growing TikaParser going forward, it is a simple start but there is definitely a space for further enhancements. Will demo the MBRs with the quick starts PR later on, thanks

sberyozkin commented:
@gsmet FYI, for the moment I'm thinking TikaParser should not be an injectable interface, I'd like to iterate few times to settle on the good enough and stable API worthy of the interface

sberyozkin commented:
And yeah, I squashed a 2nd commit :-)

agiertli commented on extensions/tika/deployment/src/main/java/io/quarkus/tika/deployment/TikaProcessor.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @BuildStep
+    public void producePdfBoxResources(BuildProducer<SubstrateResourceBuildItem> resource) throws Exception {
```
Not sure if it's required for Tika extension, but when I was trying to make PDFBox work in quarkus, when I was calling:
```
        PDFTextStripper stripper = new PDFTextStripper();
```
I was receiving this error inside native image:

https://gist.github.com/agiertli/35170cfdd5a8b3fd7e86fc5d6e238e3a

I solved it by adding Identity-h to the native-image resources:

https://github.com/cvt-oss/fb-invoice-pdf-analyzer/blob/quarkus/pom.xml#L112
Dur

sberyozkin replied:
@agiertli thanks for the review and the hint. I don't recall right now but without those (or one of those) inclusions I was getting some native error as well.
It depends on whether the `PDFTextStripper` is used on the Tika path somewhere or not. We may need to take care of the `Identity-HS` resources as well.

Can you consider giving this PR a try and see if it works for your case ? We may get some valuable early review this way, thanks

emmanuelbernard replied:
It;s better to have the extension add these resource files instead of forcing the user to set them in the pom.ml for info. That's one the reason they exist, hide the GraalVM ugliness.

sberyozkin replied:
@emmanuelbernard I agree :-) but @agiertli  has not used Tika but PDFBox directly so I'm not sure that is needed for the Tika code path yet

sberyozkin replied:
**Update**: `Identity-HS` is [here](https://github.com/apache/pdfbox/tree/trunk/fontbox/src/main/resources/org/apache/fontbox/cmap). It is hard right now to get all the resources which will actually be needed so I propose to add them whenever the exception is raised  at the Tika path.

agiertli replied:
I confirm that parsing the same PDF using tika-extension API succeed:
```
      TikaParser parser = new TikaParser();
      String invoiceText = parser.getContent(is).getText();
```

This extension also simplifies the project configuration a lot! **No more hacks** like this are needed anymore:
https://github.com/gunnarmorling/quarkus-pdf-extract/blob/master/pom.xml#L113

The only difference in behavior is that  the above code produces string with extra \n lines (while using original PDFBox API does not do this)

sberyozkin replied:
Hi @agiertli thanks, `Identity-HS` and friends may still show up on some Tika code paths for some complex PDF files :-), but we will take care of those resources if needed for sure

emmanuelbernard replied:
For info in Hobernate ORM we had a long list of files ot add and we used the following script that we run one a new version of Hibernate is integrated and adapt the processor accordingly
https://github.com/quarkusio/quarkus/blob/master/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/HqlNodeScannerTestCase.java

Might be an inspiration

sberyozkin replied:
Hi Emmanuel, yes, something like that would be needed to smartly add all the resources that may be needed (as far as I understand the idea, based on the link, would be to iterate over all, in this case, non-class resources and add them)

emmanuelbernard replied:
In practice, we generated that list and copied it in our extension code. So that we don't have to iterate over this list every single build. But that's a burden when the extension updates to the next version of the underlying library to make sure to rerun that script. It happens relatively rarely though

sberyozkin commented:
@gsmet Hi, so how does this PR look to you (and the team) ? IMHO it can be useful, will require more work to make it really useful, but until it makes it as an extension it won't be happening as fast as I'd like to as I can't predict how users will want to use it beyond the straightforward text and metadata extraction :-), even though I have a plan how to grow it in the short term.

gsmet commented:
@sberyozkin I agree it's useful, I just didn't have the time to review it yet (preparing a talk).

Not sure yet if it will make 0.18.0 as I plan to release it tomorrow (we want the GraalVM 19 support out) but it will be in 0.19.0 for sure.

sberyozkin commented:
@gsmet thanks. I agree there is no rush around having this feature in, will be happy to see it in whenever it makes it (might tweak few bits in meantime)

sberyozkin commented:
Recording here future enhancement ideas after the conversation with @agiertli : 1) optionally strip off the empty lines - will require replacing ToTextContentHandler with a custom one 2) Supper the user provided content handlers 3) add a `TikaParser.getText()` shortcut as an alternative to `TikaParser.getContent().getText()` which is simpler when no metadata is needed. 4) support the Tika parsers and parsercontexts alternative to the AutoDetect parser which can be more optimized to deal with a given file 5) add few handy utilities like check if the given PDF has been created on this date or by that author, etc

(Will open the issues after the extension makes it to the master)

gsmet commented on extensions/tika/deployment/pom.xml:
```diff
@@ -0,0 +1,74 @@
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-tika-deployment</artifactId>
+    <name>Quarkus - Tika - Deployment</name>
```
Here, we should use the full name `Apache Tika`. Same for the other artifacts.

gsmet commented on integration-tests/tika/pom.xml:
```diff
@@ -0,0 +1,135 @@
+
+    <artifactId>quarkus-integration-test-tika</artifactId>
+    <name>Quarkus - Integration Tests - Tika</name>
+    <description>The Tika integration tests module</description>
```
Same here, let's use the full name in name and description.

gsmet commented on extensions/tika/runtime/src/main/java/io/quarkus/tika/Content.java:
```diff
@@ -0,0 +1,20 @@
+        this.metadata = metadata;
+    }
+
+    public String getText() {
```
The getter should be consistent.

sberyozkin replied:
@gsmet can you clarify it please ?

gsmet replied:
The getter is called `getText()` whereas the field is called `content`.

gsmet commented on extensions/tika/runtime/src/main/java/io/quarkus/tika/Content.java:
```diff
@@ -0,0 +1,20 @@
+package io.quarkus.tika;
+
+public class Content {
```
I would name this one `TikaResult`.

sberyozkin replied:
I'm really not too keen to change `Content` to `TikaResult`, so now I'd have `parser.getResult(InputStream)` instead of `parser.getContent(InputStream)` the latter reflects better what the user is after IMHO... It looks natural to me to ask the parser to return a composite `Content` from a provided input stream

gsmet commented on extensions/tika/runtime/src/main/java/io/quarkus/tika/Metadata.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Map;
+import java.util.Set;
+
+public class Metadata {
```
I would get rid of that one and expose `TikaMetadata`. I don't see any value in having an additional abstraction layer.

sberyozkin replied:
Originally I started with this one I think, but Tika [Metadata](https://github.com/apache/tika/blob/master/tika-core/src/main/java/org/apache/tika/metadata/Metadata.java) interface is a bit verbose and somewhat complex (arrays, etc), and it is **publicly modifiable** so my plan was expose it via a simple API sufficient to get the String metadata properties.
Does it sound reasonable ?

gsmet replied:
I still think that this one should be named `TikaMetadata` to be consistent with `TikaContent`.

As for being reasonable, yes. I'm not a big fan of instantiating new objects in a getter but maybe it makes sense to not do it for all the extracted metadata, considering you would probably only use part of it.

gsmet commented on extensions/tika/runtime/src/main/java/io/quarkus/tika/TikaParser.java:
```diff
@@ -0,0 +1,72 @@
+                return new Content(tikaHandler == null ? null : tikaHandler.toString().trim(), convert(tikaMetadata));
+            }
+        } catch (Exception e) {
+            LOG.warnf("%s stream can not be parsed", contentType);
```
That's not how it should be done:
- you shouldn't have a warning
- you should have a proper message for your exception
- not sure if the `contentType` information is useful but in any case it might be an issue as if it's null, you will end up with `null stream can not be parsed` which is misleading. `Unable to parse stream for content-type ...` might be better with the last part conditioned to the `contentType` variable not being null.

sberyozkin replied:
content-type will never be null at the MBR level, it is guaranteed by the JAX-RS spec, but your proposed message reads better, happy to update to it

gsmet replied:
You're passing `null` as a `contentType` just a few lines above.

sberyozkin replied:
Sorry, I forgot it was now not hidden any longer inside MBR :-)

gsmet commented on extensions/tika/runtime/src/main/java/io/quarkus/tika/TikaParser.java:
```diff
@@ -0,0 +1,72 @@
+import org.xml.sax.ContentHandler;
+import org.xml.sax.helpers.DefaultHandler;
+
+public class TikaParser {
```
I think this should be an `@ApplicationScoped` bean that we can inject.

The Tika parser is supposed to be thread safe so it would be highly beneficial if it was instantiated only once.

gsmet reviewed (changes requested):
Hi @sberyozkin ,

Finally reviewing this, sorry for the delay.

I added a couple of comments inline, let's try to get this in for the next version!

sberyozkin commented:
Hi @gsmet thanks for the review, let me work on it... 
FYI, I'm really not too keen to change `Content` to `TikaResult`, so now I'd have `parser.getResult(InputStream)` instead of `parser.getContent(InputStream)` the latter reflects better what the user is after IMHO...

gsmet commented:
@sberyozkin maybe we could rename the method to `parseˋ? I think that would be a better choice.

sberyozkin commented:
@gsmet Changing it to `parse` is good so then `Content` would become less important to me :-) (sorry for moving the comment around)

sberyozkin commented:
Hi @gsmet 
I've updated the PR, I've only stopped from removing the Metadata wrapper for the reasons described above, let me know please if you are OK with it (IMHO it would be a bit cleaner to go with the wrapper). Also added a TikaParser.getText() shortcut as planned earlier (to support the cases when only text is needed) 

thanks

sberyozkin commented:
@gsmet I'm having recurring doubts about the `Content` to `TikaResult` renaming. 
This reads OK:
```
TikaResult result = parser.parse(inputStream);
```

but if I demo it done at the `MessageBodyReader` level (which was agreed to move out of this PR) then this is not so cool as `TikaResult` is somewhat out of context:

```
@POST
@Consumes("application/pdf")
public Response post(TikaResult result) {
}
```

I wonder if `TikaContent` could become a good compromise between the original `Content` and the current `TikaResult`:

```
TikaContent tikaContent = parser.parse(inputStream);
processTextAndMetadata(tikaContent.getText(), tikaContent.getMetadata());
```
and

```
@POST
@Consumes("application/pdf")
public Response post(TikaContent tikaContent) {
    processTextAndMetadata(tikaContent.getText(), tikaContent.getMetadata());
}
```

If `TikaResult` stays then it won't be really a problem that I'm probably exaggerating a bit :-), but i'd like to check what you think

gsmet commented:
@sberyozkin I agree with your proposal. Can you make the changes?

gsmet commented on integration-tests/tika/pom.xml:
```diff
@@ -0,0 +1,135 @@
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
```
Can you remove the copyright headers? We decided to get rid of all of them (in Java and XML files altogether).

sberyozkin commented:
@gsmet sorry, missed your comments, let me take of them...

sberyozkin commented:
@gsmet done

sberyozkin commented:
@gsmet I realized I forgot to have `TikaProcessor` to have a `FeatureBuildItem`. That should really be it :-). I have few other minor enhancements in mind but I'd rather follow up with them once this PR is merged

gsmet commented on integration-tests/tika/src/main/java/io/quarkus/it/tika/GreetingResource.java:
```diff
@@ -0,0 +1,35 @@
+import io.quarkus.tika.TikaParser;
+
+@Path("/parse")
+public class GreetingResource {
```
Could we rename this class to something that is more relevant?

gsmet commented on integration-tests/tika/src/test/java/io/quarkus/it/tika/GreetingResourceTest.java:
```diff
@@ -0,0 +1,84 @@
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+public class GreetingResourceTest {
```
Same here, the name is not consistent with what it tests.

gsmet commented on integration-tests/tika/src/test/java/io/quarkus/it/tika/NativeGreetingResourceIT.java:
```diff
@@ -0,0 +1,7 @@
+import io.quarkus.test.junit.SubstrateTest;
+
+@SubstrateTest
+public class NativeGreetingResourceIT extends GreetingResourceTest {
```
Same here about the name.

gsmet reviewed (commented):
@sberyozkin I added some hopefully final comments :).

Would you mind taking a look? Once it's done, I think we can merge this first iteration.

The next step will probably be to write a quickstart and a guide.

BTW, would you mind squashing everything in one atomic commit? Thanks!

sberyozkin commented:
@gsmet sure, renaming the `Metadata` is fine, as all other classes there are starting with `Tika` now, but what did you mean about the instantiation part ?

gsmet commented:
@sberyozkin yes, basically, every time you call `getNames()`, for instance, you will create a new wrapper.

For the keys, I think I would maybe create the list in the constructor. For the values, I'm not totally sure it's a good idea as we would create potentially a lot of wrappers, just to extract a property.

So for now, I would say let's leave it at that and if someone complains with a proper use case, we can revisit it.

sberyozkin commented:
@gsmet OK, I've also removed the wrapper on the `getSingleValue` path. Now the most difficult part, the squashing :-), I should've done straight away but thought I'd get the latest changes out asap

gsmet commented:
Following our discussion on Zulip, I rebased and force-pushed.

Let's wait for CI. Thanks for your patience and your efforts!

sberyozkin commented:
@gsmet thanks for helping to make this extension a noteworthy feature :-). CI is green now and it is good to go in. I'll need to follow up with some interesting enough quick start example which in itself will likely drive more feature enhancements)
